### ISSUE TITLE
Automated builds hang when prompt_yn() asks a question

### ISSUE THREAD

**Issue by bigpresh** [rjbs/Dist-Zilla#514]
I raised https://rt.cpan.org/Ticket/Display.html?id=111943 against Dist::Zilla::Plugin::MatchManifest, but I'm not sure that it should be worked around in the plugin - it seems more like something in D::Z that should be fixed once for all plugins?

The plugin in question asks whether to update the MANIFEST by calling `$self->zilla->chrome->prompt_yn()` - but even when we're running via `dzil smoke` (and therefore with nobody at a terminal ready to answer questions) it hangs forever waiting for a response until the build times out.

I'd be happy to supply a patch to Dist::Zilla::Chrome::Term that would make `prompt_yn()` return undef (to signify "nobody to ask, sorry") if we were running with `$ENV{AUTOMATED_TESTING}` set (or default to "n" to reduce the likelihood of unexpected breakage in existing plugins, or even allow it to take a default, which it would return in that case), if it's agreed that this is sensible and it's likely to be accepted.

**Comment by dolmen** [comment]
:+1:

But another way to fix the issue would be to use a different Dist::Zilla::Role::Chrome implementation when either `${AUTOMATED_TESTING}` or `STDIN` is not a tty (`! -t`).

**Comment by bigpresh** [comment]
Very true - adding a new ` Dist::Zilla::Chrome::NoTTY` or similar, which implements prompt_yn() to just return "n" (or whatever is decided is appropriate) may be the cleanest way.

**Comment by karenetheridge** [comment]
You (the plugin) still needs to have appropriate logic to decide when to
not prompt.

Take a look at how I handle non-interactive builds in
Dist::Zilla::Plugin::PromptIfStale. As you might guess from the name, its
entire raison d'être is prompting in the middle of the build.

**Comment by bigpresh** [comment]
> You (the plugin) still needs to have appropriate logic to decide when to not prompt.

This is the part I was wondering about - should it be each individual plugin's responsibility to know when to prompt and when not to, or should they be able to expect that D::Z will work that out for them?

**Comment by madsen** [comment]
> You (the plugin) still needs to have appropriate logic to decide when to not prompt.

Not necessarily.  I think it's perfectly reasonable for `$self->zilla->chrome->prompt_yn()` to return the default answer if it knows there's nobody to answer the question.  In MatchManifest's case, that's all it would need to do the right thing.

Plugins would only need special logic if they wanted to do something other than the default response when the build is non-interactive.

**Comment by dolmen** [comment]
> Plugins would only need special logic if they wanted to do something other than the default response when the build is non-interactive.

This implies that plugins would be able to detect if the build is interactive. But the `Dist::Zilla::Role::Chrome` does not provide that information. I think that this is where the information should be provided to plugins.

**Comment by karenetheridge** [comment]
> This implies that plugins would be able to detect if the build is interactive.

It's determined from the state of various filehandles.  The chromes are no more capable of figuring this out than the plugins themselves are.

https://metacpan.org/source/ETHER/Dist-Zilla-Plugin-PromptIfStale-0.047/lib/Dist/Zilla/Plugin/PromptIfStale.pm#L251-252

**Comment by dolmen** [comment]
> The chromes are no more capable of figuring this out than the plugins themselves are.

My point is that detection of interactive mode should be the responsability of the framework, not of each plugin.

**Comment by bigpresh** [comment]
On Sat, 27 Feb 2016 12:24:35 -0800
Olivier Mengué <notifications@github.com> wrote:

> > The chromes are no more capable of figuring this out than the
> > plugins themselves are.  
> 
> My point is that detection of interactive mode should be the
> responsability of the framework, not of each plugin.

I would tend to agree - it would seem sane for plugins to be able to
expect Dist::Zilla to select an appropriate chrome which knows if it
can ask the user questions or not.

To avoid unexpected surprises, it would perhaps be reasonable to die
unless the plugin passes a "this is the default to return if there's no
user to ask" value, though - just defaulting to "no" might be a
surprise to existing plugins.

**Comment by rjbs** [comment]
I am in favor of a non-interactive Chrome, along with autodetection to pick which chrome to use.

**Comment by rjbs** [comment]
I pushed up b63c441a.  The term chrome was already handling non-tty terminals for press-any-key.  I expanded it for y/n.

Thoughts?

**Comment by rjbs** [comment]
This was fixed in 6.010
